File: inst/scripts/fdarm-ch03.R
###
###
### Ramsey, Hooker & Graves (2009)
### Functional Data Analysis with R and Matlab (Springer)
###
### ch. 3.  How to specify basis systems for building functions
###
library(fda)
##
## Section 3.1 Basis Function Systems for Constructing Functions
##
const.basis   = create.constant.basis(rangeval=0:1)
monom.basis   = create.monomial.basis(rangeval=0:1, nbasis=1)
fourier.basis = create.fourier.basis(rangeval=0:1,
                      nbasis=5, period=1)
bspline.basis = create.bspline.basis(rangeval=0:1,
             nbasis=5, norder=2, breaks=seq(0, 1, length=5) )

##
## Section 3.2 Fourier Series for Periodic Data and Functions
##
daybasis65 = create.fourier.basis(c(0,365), 65)

T.         = 500
daybasis.T = create.fourier.basis(c(0,365), 3, period=T.)

zerobasis  = create.fourier.basis(c(0, 365), 65, dropind=1)
zerobasis. = daybasis65[2:65]
all.equal(zerobasis, zerobasis.)

fourier.basis3 = create.fourier.basis(rangeval=c(0, 1), nbasis=3)

help(create.fourier.basis)

##
## Section 3.3 Spline series for Non-periodic Data and Functions
##

#  section 3.3.3 Examples
bspline4 = create.bspline.basis(breaks=c(0, .5, 1))
knots(bspline4, interior=FALSE)

bspline2 = create.bspline.basis(breaks=c(0, .5, 1), norder=2)
knots(bspline2, interior=FALSE)

bspline2.2 = create.bspline.basis(breaks=c(0, .5, .5, 1), norder=2)
knots(bspline2.2, interior=FALSE)

bspline4.3 = create.bspline.basis(breaks=c(0, .5, .5, .5, 1))
knots(bspline4.3, interior=FALSE)

#  section 3.3.4 B-Splines
splinebasis = create.bspline.basis(c(0,10), 13)

# Figure 3.1
plot(splinebasis, xlab='t', ylab='Bspline basis functions B(t)', las=1)

# Figure 3.2
basis2 = create.bspline.basis(c(0,2*pi), 5, 2)
basis3 = create.bspline.basis(c(0,2*pi), 6, 3)
basis4 = create.bspline.basis(c(0,2*pi), 7, 4)

theta    = seq(0, 2*pi, length=201)
sin.theta= sin(theta)

sin2 = Data2fd(theta, sin.theta, basis2)
sin3 = Data2fd(theta, sin.theta, basis3)
sin4 = Data2fd(theta, sin.theta, basis4)

sin2.theta = predict(sin2, theta)
sin3.theta = predict(sin3, theta)
sin4.theta = predict(sin4, theta)

sinRng = range(sin2.theta)
pi3    = ((1:3)*pi/2)

op = par(mfrow=c(3,2), mar=c(3,4,2,2)+.1)

plot(theta, sin2.theta, type='l', ylim=sinRng, xlab='', ylab='Order = 2',
     main='sine(t)' )
lines(theta, sin.theta, lty='dashed')
abline(v=pi3, lty='dotted')

Dsin2.theta = predict(sin2, theta, 1)
plot(theta, Dsin2.theta, type='l', ylim=sinRng, xlab='', ylab='',
     main='D sine(t)')
lines(theta, cos(theta), lty='dashed')
abline(v=pi3, lty='dotted')

plot(theta, sin3.theta, type='l', ylim=sinRng, xlab='', ylab='Order = 3')
lines(theta, sin.theta, lty='dashed')
abline(v=pi3, lty='dotted')

Dsin3.theta = predict(sin3, theta, 1)
plot(theta, Dsin3.theta, type='l', ylim=sinRng, xlab='', ylab='')
lines(theta, cos(theta), lty='dashed')
abline(v=pi3, lty='dotted')

plot(theta, sin4.theta, type='l', ylim=sinRng, xlab='t', ylab='Order = 4')
lines(theta, sin.theta, lty='dashed')
abline(v=pi3, lty='dotted')

Dsin4.theta <- predict(sin4, theta, 1)
plot(theta, Dsin4.theta, type='l', ylim=sinRng, xlab='t', ylab='')
lines(theta, cos(theta), lty='dashed')
abline(v=pi3, lty='dotted')

par(op)

splinebasis6 = create.bspline.basis(c(0,10), 15, 6)

# Knots not equally spaced
spline.unequal.knots = create.bspline.basis(breaks=c(0, .7, 1))
knots(spline.unequal.knots, interior=FALSE)

plot(splinebasis6[7])
plot(splinebasis6[1:2])

t10 = seq(0, 10, .1)
spline6.10 = predict(splinebasis6, t10)
plot(t10, rowSums(spline6.10))

heightbasis  = with(growth, create.bspline.basis(c(1, 18), 35,6,age))
# or
heightbasis. = create.bspline.basis(norder=6, breaks=growth$age)
all.equal(heightbasis, heightbasis.)

help(create.bspline.basis)

##
## Section 3.4 Constant, Monomial and Other Bases
##
conbasis = create.constant.basis(c(0,1))
conb     = create.constant.basis()
all.equal(conbasis, conb)

monbasis = create.monomial.basis(c(0,1), 4)
monb     = create.monomial.basis(nbasis=4)
all.equal(monbasis, monb)

##
## Section 3.5 Methods for Functional Basis Objects
##
methods(class='basisfd')

print(monb)
print.basisfd(monb) # the same

summary(monb)

monbasis==monb

is(monb, 'basisfd')
inherits(monb, 'basisfd')

monb$params
monb$params = 2*(0:3)

t.1       = seq(0, 1, .1)
monbmat.1 = eval.basis(t.1, monb)
Dmonbmat.1= eval.basis(t.1, monb, 1)

monbmat.1. = predict(monb, t.1)
all.equal(monbmat.1, monbmat.1.)

Dmonbmat.1.= predict(monb, t.1, 1)
all.equal(Dmonbmat.1, Dmonbmat.1.)

##
## Section 3.6 The Structure of the basisfd or basis Class
##

help(basisfd)

##
## Section 3.7 Some Things to Try
##
# (exercises for the reader)

File: inst/scripts/fdarm-ch03.m
%
%
% Ramsay, Hooker & Graves (2009)
% Functional Data Analysis with R and Matlab (Springer)
%

%  Remarks and disclaimers

%  These R commands are either those in this book, or designed to 
%  otherwise illustrate how R can be used in the analysis of functional
%  data.  
%  We do not claim to reproduce the results in the book exactly by these 
%  commands for various reasons, including:
%    -- the analyses used to produce the book may not have been
%       entirely correct, possibly due to coding and accuracy issues
%       in the functions themselves 
%    -- we may have changed our minds about how these analyses should be 
%       done since, and we want to suggest better ways
%    -- the R language changes with each release of the base system, and
%       certainly the functional data analysis functions change as well
%    -- we might choose to offer new analyses from time to time by 
%       augmenting those in the book
%    -- many illustrations in the book were produced using Matlab, which
%       inevitably can imply slightly different results and graphical
%       displays
%    -- we may have changed our minds about variable names.  For example,
%       we now prefer "yearRng" to "dayrange" for the weather data.
%    -- three of us wrote the book, and the person preparing these scripts
%       might not be the person who wrote the text
%  Moreover, we expect to augment and modify these command scripts from time
%  to time as we get new data illustrating new things, add functionality
%  to the package, or just for fun.

%
% ch. 3.  How to specify basis systems for building functions
%

%  Set up some strings for constructing paths to folders.
%  These strings should be modified so as to provided access
%  to the specified folders on your computer.

%  Path to the folder containing the Matlab functional data analysis
%  software

fdaMPath = 'c:/Program Files/MATLAB/R2009a/fdaM';

addpath(fdaMPath)

%  Path to the folder containing the examples

examplesPath = [fdaMPath,'/examples'];

addpath(examplesPath)

%
% Section 3.1 Basis Function Systems for Constructing Functions
%

unitRng = [0,1];

const_basis   = create_constant_basis(unitRng);

nbasis = 1;
monom_basis   = create_monomial_basis(unitRng, nbasis);

nbasis = 5;
period = 1;
fourier_basis = create_fourier_basis(unitRng, nbasis, period);

nbasis = 5;
norder = 2;
breaks = linspace(0, 1, 5);
bspline_basis = create_bspline_basis(unitRng, nbasis, norder, breaks);

%
% Section 3.2 Fourier Series for Periodic Data and Functions
%

yearRng = [0,365];
nbasis  = 65;
daybasis65 = create_fourier_basis(yearRng, nbasis);

nbasis = 3;
period = 500;
daybasis_T = create_fourier_basis(yearRng, nbasis, period);

dropind = 1;
nbasis  = 65;
zerobasis1 = create_fourier_basis(yearRng, nbasis, dropind);
zerobasis2 = daybasis65(2:65);

nbasis = 3;
fourier_basis3 = create_fourier_basis(unitRng, nbasis);

help(create_fourier_basis)

%
% Section 3.3 Spline series for Non-periodic Data and Functions
%

%  section 3.3.3 Examples

%  order 4 spline, one interior knot

breaks = [0, .5, 1];
norder = 4;
nbasis = norder + length(breaks) - 2;
bspline4 = create_bspline_basis(unitRng, nbasis, norder, breaks);

plot(bspline4)

%  order 2 spline, one interiot knot

breaks = [0, .5, 1];
norder = 2;
nbasis = norder + length(breaks) - 2;
bspline2 = create_bspline_basis(unitRng, nbasis, norder, breaks);

plot(bspline2)

%  order 2 spline,  2 equal interior knots

breaks = [0, .5, .5, 1];
norder = 2;
nbasis = norder + length(breaks) - 2;
bspline2_2 = create_bspline_basis(unitRng, nbasis, norder, breaks);

plot(bspline2_2)

%  order 4 spline, 3 equal interior knots

breaks = [0, .5, .5, .5, 1];
norder = 2;
nbasis = norder + length(breaks) - 2;
bspline4_3 = create_bspline_basis(unitRng, nbasis, norder, breaks);

plot(bspline4_3)

%  section 3.3.4 B-Splines

splinebasis = create_bspline_basis([0,10], 13);

% Figure 3.1

plot(splinebasis) 

% Figure 3.2

twopiRng = [0,2*pi];
basis2 = create_bspline_basis(twopiRng, 5, 2);
basis3 = create_bspline_basis(twopiRng, 6, 3);
basis4 = create_bspline_basis(twopiRng, 7, 4);

theta     = linspace(0, 2*pi, 201);
sin_theta = sin(theta);

sin2 = smooth_basis(theta, sin_theta, basis2);
sin3 = smooth_basis(theta, sin_theta, basis3);
sin4 = smooth_basis(theta, sin_theta, basis4);

sin2_theta = eval_fd(sin2, theta);
sin3_theta = eval_fd(sin3, theta);
sin4_theta = eval_fd(sin4, theta);

sinRng = [min(sin2_theta), max(sin2_theta)];
pi3    = (1:3)*pi/2;

subplot(3,2,1)
plot(theta, sin2_theta, 'b-', theta, sin_theta, 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 2')
title('sine(t)')
axis([twopiRng,sinRng])

Dsin2_theta = eval_fd(sin2, theta, 1);
subplot(3,2,2)
plot(theta, Dsin2_theta, 'b-', theta, cos(theta), 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 2')
title('D sine(t)')
axis([twopiRng,sinRng])

subplot(3,2,3)
plot(theta, sin3_theta, 'b-', theta, sin_theta, 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 3')
axis([twopiRng,sinRng])

Dsin3_theta = eval_fd(sin3, theta, 1);
subplot(3,2,4)
plot(theta, Dsin3_theta, 'b-', theta, cos(theta), 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 3')
axis([twopiRng,sinRng])

subplot(3,2,5)
plot(theta, sin4_theta, 'b-', theta, sin_theta, 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 4')
axis([twopiRng,sinRng])

Dsin4_theta = eval_fd(sin4, theta, 1);
subplot(3,2,6)
plot(theta, Dsin4_theta, 'b-', theta, cos(theta), 'b--')
hold on
for i=1:3
    plot([pi3(i), pi3(i)], sinRng, 'b:') 
end
hold off
xlabel('') 
ylabel('Order = 4')
axis([twopiRng,sinRng])

%  order 6 spline with equally spaced knots

splinebasis6 = create_bspline_basis([0,10], 15, 6);

%  plot a central basis function

subplot(1,1,1)
plot(splinebasis6(7))

%  plot two left boundary basis functions

plot(splinebasis6(1:2))

%  order 4 spline with knots not equally spaced

breaks=[0, .7, 1];
norder = 4;
nbasis = norder + length(breaks) - 2;
spline_unequal_knots = ...
    create_bspline_basis(unitRng, nbasis, norder, breaks);

%  plot sum of basis function values (all equal to 1)

t10 = 0:0.1:10;
spline6_10 = eval_basis(t10, splinebasis6);
plot(t10, sum(spline6_10,2))

%
% Section 3.4 Constant, Monomial and Other Bases
%

conbasis = create_constant_basis(unitRng);
conb     = create_constant_basis();
disp(eq(conbasis, conb))

nbasis = 4;
monbasis = create_monomial_basis(unitRng, nbasis);

%
% Section 3.5 Methods for Functional Basis Objects
%

disp(monbasis)

disp(monbasis == monbasis)

isa_basis(monbasis)

disp(getbasispar(monbasis))

t_1        = 0:0.1:1;
monbmat_1  = eval_basis(t_1, monbasis);
Dmonbmat_1 = eval_basis(t_1, monbasis, 1);

%
% Section 3.6 The Structure of the basisfd or basis Class
%

help(basis)

%
% Section 3.7 Some Things to Try
%
% (exercises for the reader)
